Log attr fix

felixjulianheitmann opened the pull request:
This is my proposal for a fix of the attributes struct.

It enables a constructor for all integral/floating point types that uses std::to_string, because that is what the spec defines std::to_string for.
It uses a trivial conversion for bools.
It converts all string-like types to strings.

It is C++11 compatible

Commit:
experimenting with attributes

bc-tessonics committed:
some bug fixes to attribute fork to get compilation working

bc-tessonics committed:
attributes passed to root API

bc-tessonics committed:
forgot the things

bc-tessonics committed:
formats log messages (at least for default)

bc-tessonics committed:
more cleanup of unneeded code, needed to revert cmake standard to get this to work

bc-tessonics committed:
attribute example for testing purposes

bc-tessonics committed:
attributes can simulate structured logging, show example. Now works with pattern_formatter

bc-tessonics committed:
multiple kv pairs can be passed via initializer list

bc-tessonics committed:
fixed bugs with default formatting

bc-tessonics committed:
scrambling key and value to escape ascii codes

bc-tessonics committed:
more well-defined definition of list instantiation using a pre-defined type

bc-tessonics committed:
C++11 backwards compat fixes by replacing std::string_view with built-in string_view_t

bc-tessonics committed:
more endpoints, removed commas from logfmt

bc-tessonics committed:
code breaks when fmt is external, so am using fmts implementation only

bc-tessonics committed:
fixed errors caused by -Werror=conversion

bc-tessonics committed:
experimenting with attribute formatting

bc-tessonics committed:
actually for sure fixed the -Werror=conversion, for the specific version of gcc

bc-tessonics committed:
another pedantic error in ci

bc-tessonics committed:
custom pattern formatting is now functional

bc-tessonics committed:
fixed reordering error

bc-tessonics committed:
changed switch statement due to an interesting bug I found

bc-tessonics committed:
default kv pairs

bc-tessonics committed:
fixed example to include new pattern formatting

bc-tessonics committed:
keep old range-based for loop for non-attribute code

bachittle committed:
contextual logger attributes (#2)

* init fixed attribute, currently having issues with the factory

* using the much simpler api already defined: no factory, just define the sinks yourself

* async logger fixed attribute support

* fixed missing fixed attributes in default log calls

* forgot some spots, all cases were found?

* potential fix for illegal vector iterators

* kept up to date with attributes branch

* cleaner fixed logger api

* less function overloads, just add and remove context when needed

* fixed example to use new api

* added append function for nested context

* pushing and popping contexts for nested loggers

* fixed error due to implicit conversion

* fixed sign conversion error

* better attribute example with optional json example

Co-authored-by: Bailey Chittle <baileyc@tessonics.com>

bc-tessonics committed:
broke the utf8 code by accident

bc-tessonics committed:
global contextual logger support

bachittle committed:
attribute testing (#3)

* initial testing of attributes

* updating test cases for new api

* more testing, fixing bugs found

* fixed unused error

* errors might be caused due to passing pointers by reference, these simple pointer passes should be by value

* my only other idea for getting this to work. will experiment

* fixing tests before merge (#4)

* my only other idea for getting this to work. will experiment

* I think I fixed the thread issue, lets see...

* unused lambda capture acting funny

* see if its the context that is causing issues

* unused warning... again

* reverting to basic, see where the issue is in github actions (testing most of it locally)

* this works again?

* some diff cleanup before squash merging

* cleaner search for stop iterator

bc-tessonics committed:
exception safety tests

bc-tessonics committed:
custom tessonics readme

Commit:
fix: attr<T> constructor breaks on types that don't support std::to_string

felixjulianheitmann committed:
Re-patch log_attr: make it valid for all integral/floating types

felixjulianheitmann committed:
Re-patch log_attr: make it valid for all integral/floating types
- Is C++20 allowed?

felixjulianheitmann committed:
Make it C++11 compatible
This is more in line with spdlog

bc-tessonics committed:
overloads that work

bc-tessonics committed:
using structs to define type conversions

bc-tessonics committed:
capture every integer type

bc-tessonics committed:
forgot an overload for pure std strings

bc-tessonics committed:
matching the overloads from popular compilers

felixjulianheitmann committed:
TDS2 compatible version
Uses type traits to enable specific overloads
- std::to_string is implemented for integral types and floating point types (s. cppreference std::to_string)
- Boolean values get written as true/false
- string like types are converted to strings

felixjulianheitmann committed:
Included missing header

bc-tessonics committed:
change string view to spdlogs, remove bool type

bachittle commented:
keys and values also need to be scrambled, but when I tried it was erroring out (because it can't be called in list initalizer)

bachittle commented on include/spdlog/details/log_attr.h:
```diff
@@ -1,109 +1,42 @@
+    template<typename key_t, typename value_t, typename std::enable_if<is_string<key_t>::value, key_t>::type * = nullptr,
+        typename std::enable_if<is_string<value_t>::value, value_t>::type * = nullptr>
+    attr(key_t const &k, value_t const &v)
+        : key(std::string(k))
```
need to call scramble here, otherwise newlines show up in test cases

bachittle replied:
fixed it

bc-tessonics committed:
added scrambling

felixjulianheitmann commented:
Looks good and the tests are passing. Just notice that you now have the implicit conversion of `bool` to  `int` for `std::to_string`. This will generate `0`/`1` instead of `true`/`false` in the resulting json. Which is alright but technically incorrect.

bc-tessonics committed:
more restrictive type conversions

bc-tessonics committed:
allow shorts to be implicitly converted to int

bc-tessonics committed:
need to debug this

bc-tessonics committed:
gave up and switched back to is_integral. Most things in to_string get converted implicitly anyways...

bc-tessonics committed:
maybe still ignore booleans

bc-tessonics committed:
setting the cmake standard to 20 when using std format

bachittle committed:
setting the cmake standard to 20 when using std format (#2680)

scc-tw committed:
Fix stdout_sink_base::log's behavior inconsistency (#2646)

* Fix stdout_sink_base::log's behavior inconsistency

It will flush every time when it if not defined _WIN32, but not in
Windows family.
We viewed the commit #48d4ed9 for fixing issue #1675 .
It seems missing this flushing operation in mistake.

* Use fflush at all operating system

* Remove redundant fflush from stdout_sink_base

---------

Co-authored-by: scc <scc@teamt5.org>

bc-tessonics committed:
separate std string view from std format

bc-tessonics committed:
Merge branch 'v1.x' into seperate_std_string_view

bc-tessonics committed:
set string_view option instead of error for compat
